SourceX/controls/keymapper.hpp

```diff
@@ -0,0 +1,63 @@
+#pragma once
+
+#include <string>
+#include <functional>
+#include <memory>
+#include <vector>
+#include <unordered_map>
+
+namespace dvl {
+
+// The Keymapper maps keys to actions.
+class Keymapper final
+{
+public:
+	using SetKeyFunction = std::function<void(const std::string &key, const std::string &value)>;
+	using GetKeyFunction = std::function<std::string(const std::string &key)>;
+
+	// Action represents an action that can be triggered using a keyboard shortcut.
+	class Action final
+	{
+	public:
+		// Can this action be triggered while the player is dead?
+		enum class IfDead
+		{
+			Allow,
+			Ignore,
+		};
+
+		Action(const std::string &name, int defaultKey, std::function<void()> action):
+			name{name}, defaultKey{defaultKey}, action{action}, ifDead{IfDead::Ignore} {}
+		Action(const std::string &name, int defaultKey, std::function<void()> action, IfDead ifDead):
+			name{name}, defaultKey{defaultKey}, action{action}, ifDead{ifDead} {}
+
+		void operator()() const { action(); }
+
+	private:
+		std::string name;
+		int defaultKey;
+		std::function<void()> action;
+		IfDead ifDead;
+		int key{};
+
+		friend class Keymapper;
+	};
+
+	// Keymapper for now uses to function pointers to interact with the configuration. This is mostly a workaround and should be replaced later when another INI library will be used.
+	Keymapper(SetKeyFunction setKeyFunction, GetKeyFunction getKeyFunction);
+
+	void addAction(const Action &action);
+	void keyPressed(int key, bool isPlayerDead) const;
+	void save() const;
+	void load();
+
+private:
+	int getActionKey(const Action &action);
+
+	std::vector<std::shared_ptr<Action>> actions;
```

Comment:
There is a clear owner of actions, the `Keymapper`, so this should be `std::unique_ptr`
The map can then store `Action *` or `std::reference_wrapper<Action>`

Reply:
Yes, I thought about that. But knowing that this is C++11 only we don't have std::make_unique, which is a bit annoying. I suppose I can copy Sutter's [implementation](https://herbsutter.com/gotw/_102/) into the project though.

Reply:
Absence of `make_unique` is no excuse for gratuitous reference counting :)